### ISSUE TITLE
"tootctl cache clear" not working

### ISSUE THREAD

**Issue by maikelthedev** [mastodon/mastodon#23176]
### Steps to reproduce the problem

1. When I check an image from a user in another instance, I get the previously tooted URL of the image before that server moved to a CDN. When I go to the original page I can see the updated CDN location and therefore the image. 
2. Using tootcl cache clear should remove the old toot in MY instance so that I can reload it with the updated info. 
3. It doesn't. 
...

Please do check this conversation as it is all there: [maikelthedev@vmst.io](https://vmst.io/@maikelthedev/109722440227143649)

### Expected behaviour

cleared cache

### Actual behaviour

nothing

### Detailed description

https://vmst.io/@maikelthedev/109722440227143649

### Specifications

Masto 4.0.2

Same thing from vanilla browser UI than from Fedilab than from ELK. The problem is the cache on the visiting instance not clearing.

**Comment by smiba** [comment]
You might be mistaken on what cache clear does, as this is not referred to statuses. 

Right now there is no way to remove statuses from the database, unless they're unreferenced (not followed by any users on said instance, no user on instance has interacted with) 

Running `tootctl cache clear` just runs `Rails.cache.clear` as shown in https://github.com/mastodon/mastodon/blob/78ef635980ff391ea3ba4c37de836947a97958a0/lib/mastodon/cache_cli.rb#L15-L19

**Comment by maikelthedev** [comment]
So what does Rails cache clear do. Do I need to understand Ruby on Rails inside out for "cache clear" to mean something different than what most non-Ruby on Rails Mastodon admins expect it to do? 

Because every admin I asked thinks that described above (clearing up old statuses) is what tootctl clear cache does. 

Maybe this PR should be about "clear out the cache storage" being more specific of what it actually does, what is that "cache storage" exactly.

**Comment by smiba** [comment]
> Because every admin I asked thinks that described above (clearing up old statuses) is what tootctl clear cache does.

I missed this in my comment above, but clearing up old unreferenced statuses is done trough `tootctl statuses remove` 
(Documented in https://docs.joinmastodon.org/admin/tootctl/#statuses-remove)

Keep in mind that it only removes statuses that are entirely not interacted with, so anything as much as a reply, boost or just a favourite will keep this command from touching it.


> So what does Rails cache clear do. Do I need to understand Ruby on Rails inside out for "cache clear" to mean something different than what most non-Ruby on Rails Mastodon admins expect it to do?

As I've unfortunately never dug deep into redis I can't tell exactly for sure what cache clear does, but I'd assume it clears values stored in redis using the "Rails.cache.write" call. You can find these by searching in the code for them. 

Redis will never store anything that could be disastrous if lost (such as entire statuses)

**Comment by smiba** [comment]
> > entirely not interacted with, so anything as much as a reply, boost or just a favorite
> 
> Are we sure of this behavior? I just found out that statuses I previously boosted and fav'ed have now disappeared completely after using statuses remove. I'm the sole active user of my instance right now.

Yes, this should not happen as shown in

https://github.com/mastodon/mastodon/blob/95fb53c53edc0908334b3fa1616861809ecedd11/lib/mastodon/statuses_cli.rb#L71-L82

Personally I'd say your maintenance script is excessive, especially running accounts cull might be unwanted as this queries and thus puts (a minor amount) of load on the remote servers. Not sure how often you run these for, but a lot of these comments might be better ran less often.

Additionally cache clear is unlikely to help you much and will only slow down your instance as it has to re-generate some of the in-memory cached values stored in redis.

**Comment by pedroCX486** [comment]
> > > entirely not interacted with, so anything as much as a reply, boost or just a favorite
> > 
> > 
> > Are we sure of this behavior? I just found out that statuses I previously boosted and fav'ed have now disappeared completely after using statuses remove. I'm the sole active user of my instance right now.
> 
> Yes, this should not happen as shown in
> 
> https://github.com/mastodon/mastodon/blob/95fb53c53edc0908334b3fa1616861809ecedd11/lib/mastodon/statuses_cli.rb#L71-L82
> 
> Personally I'd say your maintenance script is excessive, especially running accounts cull might be unwanted as this does remote queries and thus puts (a minor amount) of load on the remote servers. Not sure how often you run these for, but a lot of these commands might be better ran less often.
> 
> Additionally cache clear is unlikely to help you much and will only slow down your instance as it has to re-generate some of the in-memory cached values stored in redis. Unless you're trying to save a mere few MB of memory it doesn't make much 

I only run them once every two months (but thanks for assuming I DDoS servers all around daily) to clean up messed up data, but have removed most of the lines except for the media maintenance ones.

But I still can’t help but notice that favorites, boosts and bookmarks have all disappeared.

**Comment by smiba** [comment]
> But I still can’t help but notice that favorites, boosts and bookmarks have all disappeared after the statuses remove. Guess I won’t be doing that anymore.

Has this always been happening, or only after you've updated to v4.1.0rc1? (looked for your username online and found your instance, also keep in mind you're running the glitch-soc fork)

Because if this happened after updating you might want to keep an eye on https://github.com/mastodon/mastodon/issues/23206

**Comment by pedroCX486** [comment]
> Has this always been happening, or only after you've updated to v4.1.0rc1? (looked for your username online and found your instance, but also keep in mind you're running the glitch-soc fork and not mainline mastodon)
> 
> Because if this only started happening after updating you might want to keep an eye on #23206

Happened after updating while on mainline, and yes I'm very well aware of glitch-soc (I'm not as stupid as I sound), I migrated to it on purpose when I noticed this happening thinking 4.1.0rc1 corrupted something.

**Comment by 33b5e5** [comment]
> This is what my current maintenance script looks like: <img alt="image" width="672" src="https://user-images.githubusercontent.com/44985060/214123631-d0dcde8c-8c77-4ea9-b739-c1d89cbdb183.png">

@pedroCX486 I think the first two lines of your script are redundant. From `tootctl media help remove`:

```
If --prune-profiles is specified, only avatars and headers are removed. If --remove-headers is specified, only headers are removed.
```

If you want to delete both avatars and headers, just run with `--prune-profiles`. No need for `--remove-headers` unless you ONLY want to delete headers.

**Comment by pedroCX486** [comment]
> > This is what my current maintenance script looks like: <img alt="image" width="672" src="https://user-images.githubusercontent.com/44985060/214123631-d0dcde8c-8c77-4ea9-b739-c1d89cbdb183.png">
> 
> @pedroCX486 I think the first two lines of your script are redundant. From `tootctl media help remove`:
> 
> ```
> If --prune-profiles is specified, only avatars and headers are removed. If --remove-headers is specified, only headers are removed.
> ```
> 
> If you want to delete both avatars and headers, just run with `--prune-profiles`. No need for `--remove-headers` unless you ONLY want to delete headers.

Yep! Learned this in another issue, thanks for the tip! Already removed from my script, I also removed the other settings only leaving remove-orphans and preview_cards.
